File: src/main/java/com/example/creditreportservice/service/PanValidator.java
package com.example.creditreportservice.service;

import com.example.creditreportservice.config.RegistrationProperties;
import com.example.creditreportservice.exception.PanValidationException;
import com.example.creditreportservice.service.ocr.OcrResult;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.regex.Pattern;

/**
 * Validates submitted PAN data against the OCR'd image content. Format check
 * is regex-based; image consistency check compares the submitted PAN number
 * (exact) and name (fuzzy via Levenshtein) to the OCR output.
 *
 * <p>Per the chosen design, this does <strong>not</strong> verify PAN
 * authenticity against the income-tax database — only format + OCR consistency.
 */
@Service
@RequiredArgsConstructor
public class PanValidator {

    private static final Pattern PAN_FORMAT =
            Pattern.compile("^[A-Z]{5}[0-9]{4}[A-Z]$");

    private final RegistrationProperties props;

    /** Throws {@link PanValidationException} on any failure; otherwise returns void. */
    public void validate(String submittedPan, String submittedName, OcrResult ocr) {
        validateFormat(submittedPan);
        if (submittedName == null || submittedName.isBlank()) {
            throw new PanValidationException("Full name is required");
        }
        if (ocr == null) {
            throw new PanValidationException("OCR could not read the PAN image");
        }
        if (ocr.confidence() < props.getOcr().getMinConfidence()) {
            throw new PanValidationException(
                    "PAN image was not clear enough (confidence "
                            + String.format("%.2f", ocr.confidence()) + ")");
        }

        // PAN number must match exactly after normalization.
        String ocrPan = ocr.panNumber() == null ? null : ocr.panNumber().toUpperCase().trim();
        String submitted = submittedPan.toUpperCase().trim();
        if (ocrPan == null || !ocrPan.equals(submitted)) {
            throw new PanValidationException(
                    "PAN number on the image does not match the entered value");
        }

        // Name match is fuzzy to tolerate OCR spacing / minor differences.
        if (ocr.name() == null) {
            throw new PanValidationException("Could not read name from the PAN image");
        }
        String a = normalizeName(submittedName);
        String b = normalizeName(ocr.name());
        int distance = levenshtein(a, b);
        if (distance > props.getPan().getNameMatchDistance()) {
            throw new PanValidationException(
                    "Name on the image does not match the entered name");
        }
    }

    static void validateFormat(String pan) {
        if (pan == null || !PAN_FORMAT.matcher(pan.toUpperCase().trim()).matches()) {
            throw new PanValidationException(
                    "PAN must be 5 letters, 4 digits, 1 letter (e.g. ABCDE1234F)");
        }
    }

    private static String normalizeName(String s) {
        return s == null ? ""
                : s.trim()
                        .replaceAll("[^A-Za-z\\s]", "")
                        .replaceAll("\\s+", " ")
                        .toLowerCase();
    }

    /** Classic DP Levenshtein; small strings so O(m*n) is fine. */
    static int levenshtein(String a, String b) {
        int m = a.length();
        int n = b.length();
        if (m == 0) return n;
        if (n == 0) return m;
        int[] prev = new int[n + 1];
        int[] curr = new int[n + 1];
        for (int j = 0; j <= n; j++) prev[j] = j;
        for (int i = 1; i <= m; i++) {
            curr[0] = i;
            for (int j = 1; j <= n; j++) {
                int cost = a.charAt(i - 1) == b.charAt(j - 1) ? 0 : 1;
                curr[j] = Math.min(Math.min(curr[j - 1] + 1, prev[j] + 1), prev[j - 1] + cost);
            }
            int[] tmp = prev; prev = curr; curr = tmp;
        }
        return prev[n];
    }

}

File: internal/service/pan_validator.go
package service

import (
	"fmt"
	"regexp"
	"strings"
	"unicode"

	"credit-report-service/internal/apperr"
	"credit-report-service/internal/config"
	"credit-report-service/internal/ocr"
)

var panFormat = regexp.MustCompile(`^[A-Z]{5}[0-9]{4}[A-Z]$`)

// PanValidator enforces PAN format and OCR consistency (PAN exact, name fuzzy).
// It does NOT verify PAN authenticity against the income-tax database.
type PanValidator struct {
	cfg config.PANConfig
}

func NewPanValidator(cfg config.PANConfig) *PanValidator {
	return &PanValidator{cfg: cfg}
}

// Validate throws an apperr on any failure; otherwise returns nil.
func (v *PanValidator) Validate(submittedPAN, firstName, lastName string, r *ocr.Result) error {
	if !panFormat.MatchString(strings.ToUpper(strings.TrimSpace(submittedPAN))) {
		return apperr.NewPanFailure("PAN must be 5 letters, 4 digits, 1 letter (e.g. ABCDE1234F)")
	}
	if strings.TrimSpace(firstName) == "" || strings.TrimSpace(lastName) == "" {
		return apperr.NewPanFailure("First name and last name are required")
	}
	if r == nil {
		return apperr.NewPanFailure("OCR could not read the PAN image")
	}
	if r.Confidence < v.cfg.MinConfidence {
		return apperr.NewPanFailure(fmt.Sprintf(
			"PAN image was not clear enough (confidence %.2f)", r.Confidence))
	}

	// PAN number must match exactly after normalization.
	submitted := strings.ToUpper(strings.TrimSpace(submittedPAN))
	ocrPAN := ""
	if r.PANNumber != "" {
		ocrPAN = strings.ToUpper(strings.TrimSpace(r.PANNumber))
	}
	if ocrPAN != submitted {
		return apperr.NewPanFailure("PAN number on the image does not match the entered value")
	}

	// Name match is fuzzy.
	if r.Name == "" {
		return apperr.NewPanFailure("Could not read name from the PAN image")
	}
	fullName := strings.TrimSpace(firstName + " " + lastName)
	a := normalizeName(fullName)
	b := normalizeName(r.Name)
	if levenshtein(a, b) > v.cfg.NameMatchDistance {
		return apperr.NewPanFailure("Name on the image does not match the entered name")
	}
	return nil
}

// normalizeName: lowercase, strip non-letter/non-space, collapse spaces.
func normalizeName(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	out := strings.Builder{}
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsSpace(r) {
			out.WriteRune(r)
		}
	}
	// collapse runs of whitespace
	fields := strings.Fields(out.String())
	return strings.Join(fields, " ")
}

// levenshtein is the classic O(m*n) DP edit distance.
func levenshtein(a, b string) int {
	if a == b {
		return 0
	}
	if len(a) == 0 {
		return len(b)
	}
	if len(b) == 0 {
		return len(a)
	}
	prev := make([]int, len(b)+1)
	curr := make([]int, len(b)+1)
	for j := 0; j <= len(b); j++ {
		prev[j] = j
	}
	for i := 1; i <= len(a); i++ {
		curr[0] = i
		for j := 1; j <= len(b); j++ {
			cost := 1
			if a[i-1] == b[j-1] {
				cost = 0
			}
			del := prev[j] + 1
			ins := curr[j-1] + 1
			sub := prev[j-1] + cost
			min := del
			if ins < min {
				min = ins
			}
			if sub < min {
				min = sub
			}
			curr[j] = min
		}
		prev, curr = curr, prev
	}
	return prev[len(b)]
}
